README.md
# OpenTelemetry Multiplexed Sampler for Go

[![Go Reference](https://pkg.go.dev/badge/github.com/mashiike/otelmuxsampler.svg)](https://pkg.go.dev/github.com/mashiike/otelmuxsampler)
[![Go Report Card](https://goreportcard.com/badge/github.com/mashiike/otelmuxsampler)](https://goreportcard.com/report/github.com/mashiike/otelmuxsampler)
[![MIT License](https://img.shields.io/badge/license-MIT-blue.svg)](https://opensource.org/licenses/MIT)

## Overview

**github.com/mashiike/otelmuxsampler** is a Go package for advanced sampling control in applications using OpenTelemetry's Go implementation. It supports complex sampling rules and allows you to combine them to create flexible sampling strategies. By using this package, you can effectively manage the collection of trace data and optimize your application's performance.

## Installation

You can install the **github.com/mashiike/otelmuxsampler** package using the following command:

```bash
go get github.com/mashiike/otelmuxsampler
```
## Usage

for example, you can use it like this: 
```go
package main

import (
	"context"
	"fmt"
	"log"
	"math/rand"
	"net/http"
	"os"
	"time"

	"github.com/mashiike/otelmuxsampler"
	"go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp"
	"go.opentelemetry.io/otel"
	"go.opentelemetry.io/otel/attribute"
	"go.opentelemetry.io/otel/exporters/stdout/stdouttrace"
	sdktrace "go.opentelemetry.io/otel/sdk/trace"
	"go.opentelemetry.io/otel/trace"
)

func main() {
	cleanup, err := setupTraceProvider()
	if err != nil {
		log.Fatal(err)
	}
	defer cleanup()

	mux := http.NewServeMux()
	mux.HandleFunc("/critical", func(w http.ResponseWriter, r *http.Request) {
		w.Write([]byte("critical"))
	})
	mux.HandleFunc("/high", func(w http.ResponseWriter, r *http.Request) {
		w.Write([]byte("high"))
	})
	mux.HandleFunc("/medium", func(w http.ResponseWriter, r *http.Request) {
		w.Write([]byte("medium"))
	})
	mux.HandleFunc("/low", func(w http.ResponseWriter, r *http.Request) {
		w.Write([]byte("low"))
	})
	paths := []string{"/critical", "/high", "/medium", "/low"}
	rnd := rand.New(rand.NewSource(time.Now().UnixNano()))
	mux.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
		redirectTo := paths[rnd.Intn(len(paths))]
		http.Redirect(w, r, redirectTo, http.StatusFound)
	})
	middleware := func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			log.Println("[info] request:", r.URL.Path)
			otelhttp.NewHandler(otelhttp.WithRouteTag(r.URL.Path, next), r.URL.Path, otelhttp.WithSpanOptions(
				trace.WithAttributes(attribute.String("server.request.class", getRequestClass(r))),
			)).ServeHTTP(w, r)
		})
	}
	log.Println("[info] start server, access to http://localhost:8080/")
	if err := http.ListenAndServe(":8080", middleware(mux)); err != nil {
		log.Fatal("[error]", err)
	}
}

func getRequestClass(r *http.Request) string {
	switch r.URL.Path {
	case "/critical":
		return "critical"
	case "/high":
		return "high"
	case "/medium":
		return "medium"
	default:
		return "low"
	}
}

func setupTraceProvider() (func(), error) {
	exporter, err := stdouttrace.New(
		stdouttrace.WithPrettyPrint(),
		stdouttrace.WithWriter(os.Stdout),
	)
	if err != nil {
		return func() {}, fmt.Errorf("failed to create stdout exporter: %w", err)
	}
	mux := otelmuxsampler.Multiplexed(sdktrace.NeverSample())
	mux.AttributeEqual(attribute.String("server.request.class", "critical"), sdktrace.AlwaysSample())
	mux.AttributeEqual(attribute.String("server.request.class", "high"), sdktrace.TraceIDRatioBased(0.5))
	mux.AttributeEqual(attribute.String("server.request.class", "medium"), sdktrace.TraceIDRatioBased(0.01))
	mux.AttributeEqual(attribute.String("server.request.class", "low"), sdktrace.NeverSample())

	tp := sdktrace.NewTracerProvider(
		sdktrace.WithBatcher(exporter),
		sdktrace.WithSampler(sdktrace.ParentBased(mux)),
	)
	otel.SetTracerProvider(tp)
	log.Println("[info] setup trace provider")
	return func() {
		ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
		defer cancel()
		if err := tp.Shutdown(ctx); err != nil {
			log.Println("[error] failed to shutdown:", err)
		}
	}, nil
}
```

## License

**github.com/mashiike/otelmuxsampler** is licensed under the MIT License.
See [LICENSE](./LICENSE) for the full license text.



CHANGELOG.md
# Changelog

## [v0.1.0](https://github.com/mashiike/otelmuxsampler/commits/v0.1.0) - 2023-09-03


example_test.go
package otelmuxsampler_test

import (
	"context"
	"fmt"
	"log"
	"sync"
	"time"

	"github.com/mashiike/otelmuxsampler"
	"go.opentelemetry.io/otel/attribute"
	sdktrace "go.opentelemetry.io/otel/sdk/trace"
	"go.opentelemetry.io/otel/trace"
)

type onMemoryExporter struct {
	mu    sync.Mutex
	spans []sdktrace.ReadOnlySpan
}

func (e *onMemoryExporter) ExportSpans(ctx context.Context, spans []sdktrace.ReadOnlySpan) error {
	e.mu.Lock()
	defer e.mu.Unlock()
	e.spans = append(e.spans, spans...)
	return nil
}

func (e *onMemoryExporter) Shutdown(ctx context.Context) error {
	return nil
}

func Example() {
	exporter := &onMemoryExporter{}
	// only priority=high span is sampled
	mux := otelmuxsampler.Multiplexed(sdktrace.NeverSample())
	mux.AttributeEqual(attribute.String("priority", "high"), sdktrace.AlwaysSample())
	tp := sdktrace.NewTracerProvider(
		sdktrace.WithBatcher(exporter),
		sdktrace.WithSampler(sdktrace.ParentBased(mux)),
	)
	// operation1 is priority=high, operation2 is priority=low
	tr := tp.Tracer("example")
	func() {
		_, span := tr.Start(nil, "operation1", trace.WithAttributes(
			attribute.String("priority", "high"),
		))
		defer span.End()
	}()
	func() {
		_, span := tr.Start(nil, "operation2", trace.WithAttributes(
			attribute.String("priority", "low"),
		))
		defer span.End()
	}()
	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()
	if err := tp.Shutdown(ctx); err != nil {
		log.Fatal("failed to shutdown:", err)
	}

	// stdout to span names
	for _, span := range exporter.spans {
		fmt.Println(span.Name())
	}
	// Output:
	// operation1
}


sampler_test.go
package otelmuxsampler_test

import (
	"fmt"
	"testing"

	"github.com/mashiike/otelmuxsampler"
	"go.opentelemetry.io/otel/attribute"
	sdktrace "go.opentelemetry.io/otel/sdk/trace"
	"go.opentelemetry.io/otel/trace"
)

func decisionToString(d sdktrace.SamplingDecision) string {
	switch d {
	case sdktrace.RecordAndSample:
		return "RecordAndSample"
	case sdktrace.RecordOnly:
		return "RecordOnly"
	case sdktrace.Drop:
		return "Drop"
	default:
		return "Unknown"
	}
}

func TestMultiplexerAttributeEqual(t *testing.T) {
	mux := otelmuxsampler.Multiplexed(sdktrace.NeverSample())
	mux.AttributeEqual(attribute.String("server.request.class", "critical"), sdktrace.AlwaysSample())
	mux.AttributeEqual(attribute.String("server.request.class", "high"), sdktrace.NeverSample())
	mux.AttributeEqual(attribute.String("server.request.class", "medium"), sdktrace.NeverSample())
	mux.AttributeEqual(attribute.String("server.request.class", "low"), sdktrace.NeverSample())

	traceID, _ := trace.TraceIDFromHex("4bf92f3577b34da6a3ce929d0e0e4736")

	t.Run("critical", func(t *testing.T) {
		params := sdktrace.SamplingParameters{
			TraceID: traceID,
			Attributes: []attribute.KeyValue{
				attribute.String("server.request.class", "critical"),
			},
		}
		actualDecision := mux.ShouldSample(params).Decision
		expectedDecision := sdktrace.AlwaysSample().ShouldSample(params).Decision
		if actualDecision != expectedDecision {
			t.Errorf("ShouldSample should be %s, got %s instead",
				decisionToString(expectedDecision),
				decisionToString(actualDecision),
			)
		}
	})
	t.Run("low", func(t *testing.T) {
		params := sdktrace.SamplingParameters{
			TraceID: traceID,
			Attributes: []attribute.KeyValue{
				attribute.String("server.request.class", "low"),
			},
		}
		actualDecision := mux.ShouldSample(params).Decision
		expectedDecision := sdktrace.NeverSample().ShouldSample(params).Decision
		if actualDecision != expectedDecision {
			t.Errorf("ShouldSample should be %s, got %s instead",
				decisionToString(expectedDecision),
				decisionToString(actualDecision),
			)
		}
	})
}

func TestMultiplexerDescriptoin(t *testing.T) {
	mux := otelmuxsampler.Multiplexed(sdktrace.AlwaysSample())
	mux.Handle("internal_ignore", otelmuxsampler.NewHandler(
		func(p sdktrace.SamplingParameters) bool {
			return p.Kind == trace.SpanKindInternal
		},
		sdktrace.NeverSample(),
	))
	mux.Handle("server_always", otelmuxsampler.NewHandler(
		func(p sdktrace.SamplingParameters) bool {
			return p.Kind == trace.SpanKindServer
		},
		sdktrace.AlwaysSample(),
	))
	expectedDescription := fmt.Sprintf(
		"Multiplexer{internal_ignore:%s,server_always:%s,default:%s}",
		sdktrace.NeverSample().Description(),
		sdktrace.AlwaysSample().Description(),
		sdktrace.AlwaysSample().Description(),
	)

	if mux.Description() != expectedDescription {
		t.Errorf("Sampler description should be %s, got '%s' instead",
			expectedDescription,
			mux.Description(),
		)
	}
}


sampler.go
package otelmuxsampler

import (
	"fmt"
	"strings"
	"sync"

	"go.opentelemetry.io/otel/attribute"
	sdktrace "go.opentelemetry.io/otel/sdk/trace"
)

// Multiplexer is a Sampler that multiplexes sampling decisions
type Multiplexer struct {
	mu             sync.RWMutex
	defaultSampler sdktrace.Sampler
	lastIndex      int
	entries        []entry
	description    string
}

type entry struct {
	index   int
	name    string
	handler Handler
}

// Multiplexed returns a new Multiplexer
func Multiplexed(sampler sdktrace.Sampler) *Multiplexer {
	return &Multiplexer{
		defaultSampler: sampler,
	}
}

func (mux *Multiplexer) Handle(name string, handler Handler) {
	mux.mu.Lock()
	defer mux.mu.Unlock()
	mux.lastIndex++
	mux.entries = append(mux.entries, entry{
		index:   mux.lastIndex,
		name:    name,
		handler: handler,
	})

	var builder strings.Builder
	builder.WriteString("Multiplexer{")
	for _, e := range mux.entries {
		builder.WriteString(e.name)
		builder.WriteString(":")
		builder.WriteString(e.handler.Description())
		builder.WriteString(",")
	}
	builder.WriteString("default:")
	builder.WriteString(mux.defaultSampler.Description())
	builder.WriteString("}")
	mux.description = builder.String()
}

func (mux *Multiplexer) HandleFunc(name string, matcher func(sdktrace.SamplingParameters) bool, sampler sdktrace.Sampler) {
	mux.Handle(name, NewHandler(matcher, sampler))
}

func (mux *Multiplexer) AttributeEqual(attr attribute.KeyValue, sampler sdktrace.Sampler) {
	mux.Handle(
		fmt.Sprintf("attribute_equal(%s=%s)", attr.Key, attr.Value.Emit()),
		NewHandler(func(p sdktrace.SamplingParameters) bool {
			for _, a := range p.Attributes {
				if a.Key != attr.Key {
					continue
				}
				if a.Value.Emit() == attr.Value.Emit() {
					return true
				}
			}
			return false
		}, sampler),
	)
}

func (mux *Multiplexer) AttributeExists(attrKey attribute.Key, sampler sdktrace.Sampler) {
	mux.Handle(
		fmt.Sprintf("attribute_exists(%s)", attrKey),
		NewHandler(func(p sdktrace.SamplingParameters) bool {
			for _, a := range p.Attributes {
				if a.Key == attrKey {
					return true
				}
			}
			return false
		}, sampler),
	)
}

func (mux *Multiplexer) ShouldSample(p sdktrace.SamplingParameters) sdktrace.SamplingResult {
	mux.mu.RLock()
	defer mux.mu.RUnlock()
	for _, e := range mux.entries {
		if e.handler.Match(p) {
			return e.handler.ShouldSample(p)
		}
	}
	return mux.defaultSampler.ShouldSample(p)
}

func (mux *Multiplexer) Description() string {
	mux.mu.RLock()
	defer mux.mu.RUnlock()
	return mux.description
}


example/main.go
package main

import (
	"context"
	"fmt"
	"log"
	"math/rand"
	"net/http"
	"os"
	"time"

	"github.com/mashiike/otelmuxsampler"
	"go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp"
	"go.opentelemetry.io/otel"
	"go.opentelemetry.io/otel/attribute"
	"go.opentelemetry.io/otel/exporters/stdout/stdouttrace"
	sdktrace "go.opentelemetry.io/otel/sdk/trace"
	"go.opentelemetry.io/otel/trace"
)

func main() {
	cleanup, err := setupTraceProvider()
	if err != nil {
		log.Fatal(err)
	}
	defer cleanup()

	mux := http.NewServeMux()
	mux.HandleFunc("/critical", func(w http.ResponseWriter, r *http.Request) {
		w.Write([]byte("critical"))
	})
	mux.HandleFunc("/high", func(w http.ResponseWriter, r *http.Request) {
		w.Write([]byte("high"))
	})
	mux.HandleFunc("/medium", func(w http.ResponseWriter, r *http.Request) {
		w.Write([]byte("medium"))
	})
	mux.HandleFunc("/low", func(w http.ResponseWriter, r *http.Request) {
		w.Write([]byte("low"))
	})
	paths := []string{"/critical", "/high", "/medium", "/low"}
	rnd := rand.New(rand.NewSource(time.Now().UnixNano()))
	mux.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
		redirectTo := paths[rnd.Intn(len(paths))]
		http.Redirect(w, r, redirectTo, http.StatusFound)
	})
	middleware := func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			log.Println("[info] request:", r.URL.Path)
			otelhttp.NewHandler(otelhttp.WithRouteTag(r.URL.Path, next), r.URL.Path, otelhttp.WithSpanOptions(
				trace.WithAttributes(attribute.String("server.request.class", getRequestClass(r))),
			)).ServeHTTP(w, r)
		})
	}
	log.Println("[info] start server, access to http://localhost:8080/")
	if err := http.ListenAndServe(":8080", middleware(mux)); err != nil {
		log.Fatal("[error]", err)
	}
}

func getRequestClass(r *http.Request) string {
	switch r.URL.Path {
	case "/critical":
		return "critical"
	case "/high":
		return "high"
	case "/medium":
		return "medium"
	default:
		return "low"
	}
}

func setupTraceProvider() (func(), error) {
	exporter, err := stdouttrace.New(
		stdouttrace.WithPrettyPrint(),
		stdouttrace.WithWriter(os.Stdout),
	)
	if err != nil {
		return func() {}, fmt.Errorf("failed to create stdout exporter: %w", err)
	}
	mux := otelmuxsampler.Multiplexed(sdktrace.NeverSample())
	mux.AttributeEqual(attribute.String("server.request.class", "critical"), sdktrace.AlwaysSample())
	mux.AttributeEqual(attribute.String("server.request.class", "high"), sdktrace.TraceIDRatioBased(0.5))
	mux.AttributeEqual(attribute.String("server.request.class", "medium"), sdktrace.TraceIDRatioBased(0.01))
	mux.AttributeEqual(attribute.String("server.request.class", "low"), sdktrace.NeverSample())

	tp := sdktrace.NewTracerProvider(
		sdktrace.WithBatcher(exporter),
		sdktrace.WithSampler(sdktrace.ParentBased(mux)),
	)
	otel.SetTracerProvider(tp)
	log.Println("[info] setup trace provider")
	return func() {
		ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
		defer cancel()
		if err := tp.Shutdown(ctx); err != nil {
			log.Println("[error] failed to shutdown:", err)
		}
	}, nil
}
